test/helpers/foundry/helpers.mts

```diff
@@ -0,0 +1,393 @@
+      const response = await download.response();
+      const contentLength = response.headers['content-length'];
+      return contentLength ? parseInt(contentLength, 10) : 0;
+    },
+    stream(offset: number, bytes: number) {
+      const options = {
+        agent,
+        headers: {
+          range: `bytes=${offset}-${bytes ? offset + bytes : ''}`,
+        },
+      };
+      return startDownload(url, options);
+    },
+  };
+
+  const { files } = await Unzip.custom(source);
+  const filtered = files.filter(({ path }) =>
+    binaries.includes(basename(path, extname(path))),
+  );
+  return await Promise.all(
+    filtered.map(async ({ stream, path }) => {
+      const dest = join(dir, path);
+      await pipeline(stream(), createWriteStream(dest));
+      return dest;
+    }),
+  );
+}
+
+type DownloadOptions = {
+  method?: 'GET' | 'HEAD';
+  headers?: Record<string, string>;
+  agent?: HttpsAgent | HttpAgent;
+  maxRedirects?: number;
+};
+
+/**
+ * Starts a download from the given URL.
+ *
+ * @param url The URL to download from
+ * @param options The download options
+ * @param redirects The number of redirects that have occurred
+ * @returns A stream of the download
+ */
+function startDownload(
+  url: URL,
+  options: DownloadOptions = {},
+  redirects: number = 0,
+): DownloadStream {
+  const MAX_REDIRECTS = options.maxRedirects ?? 5;
+  const request = url.protocol === 'http:' ? httpRequest : httpsRequest;
+  const stream = new DownloadStream();
+  request(url, options, async (response) => {
+    stream.once('close', () => {
+      response.destroy();
+    });
+
+    const { statusCode, statusMessage, headers } = response;
+    // handle redirects
+    if (
+      statusCode &&
+      statusCode >= 300 &&
+      statusCode < 400 &&
+      headers.location
+    ) {
+      if (redirects >= MAX_REDIRECTS) {
+        stream.emit('error', new Error('Too many redirects'));
+        response.destroy();
+      } else {
+        // note: we don't emit a response until we're done redirecting, because
+        // handlers only expect it to be emitted once.
+        await pipeline(
+          startDownload(new URL(headers.location, url), options, redirects + 1)
+            // emit the response event to the stream
+            .once('response', stream.emit.bind(stream, 'response')),
+          stream,
+        ).catch(stream.emit.bind(stream, 'error'));
+        response.destroy();
+      }
+    }
+    // check for HTTP errors
+    else if (!statusCode || statusCode < 200 || statusCode >= 300) {
+      stream.emit(
+        'error',
+        new Error(
+          `Request to ${url} failed. Status Code: ${statusCode} - ${statusMessage}`,
+        ),
+      );
+      response.destroy();
+    } else {
+      // resolve with response stream
+      stream.emit('response', response);
+
+      response.once('error', stream.emit.bind(stream, 'error'));
+      await pipeline(response, stream).catch(stream.emit.bind(stream, 'error'));
```

Comment:
More cool `pipeline` usage!

the call to `startDownload` created the `stream` earlier (a `PassThrough` steam), and returned it immediately even thought we hadn't actually downloaded anything yet. Now that we have started receiving data we just send it along.

Notice that we also emit errors on the `steam` object when appropriate.